content/learning-paths/cross-platform/simd-on-rust/simd-on-rust-part1.md

```diff
@@ -0,0 +1,348 @@
+* You will note `#[inline(never)]` in the definition of `average_vec`. This is to hint to the compiler that it should not inline this function for the same reason as previously, to help comparing against the C version.
+
+Let's check the assembly output, using `objdump -S average2` to check `average_vec` function:
+
+```asm
+00000000000069c4 <_ZN8average211average_vec17h7214a2d335bcab6cE>:
+    69c4:       4f0167e0        movi    v0.4s, #0x3f, lsl #24
+    69c8:       aa1f03e8        mov     x8, xzr
+    69cc:       3ce86821        ldr     q1, [x1, x8]
+    69d0:       3ce86842        ldr     q2, [x2, x8]
+    69d4:       4e22d421        fadd    v1.4s, v1.4s, v2.4s
+    69d8:       6e20dc21        fmul    v1.4s, v1.4s, v0.4s
+    69dc:       3ca86801        str     q1, [x0, x8]
+    69e0:       91004108        add     x8, x8, #0x10
+    69e4:       f140051f        cmp     x8, #0x1, lsl #12
+    69e8:       54ffff21        b.ne    69cc <_ZN8average211average_vec17h7214a2d335bcab6cE+0x8>  // b.any
+    69ec:       d65f03c0        ret
+```
+
+Apart from minor differences, you will note that the main loop is the same, as we expected, the 2 x `ldr` instructions followed by `fadd`, `fmul` and an `str`.
+
+### Feature detection in Rust
+
+Let's expand a bit more on the feature detection.
+
+Using SIMD intrinsics in Rust is only possible if the specific feature of the architecture that enables these intrinsics is enabled.
+
+This can only happen in architecture specific portion of the code marked by `#[cfg(target_arch = "aarch64")]`. This code will only be compiled for that particular `target_arch`, which means you might have to provide some architecture-independent implementation that will work in the other architectures if you care about your code being portable.
+
+The feature detection in particular refers to particular minor extensions in the ISA not covered by the main `target_arch` detection. For example, in the case of Aarch64, there is the `dotprod` extension that includes intrinsics such as `vdotq_u32`. The equivalent check in C would be something like:
+
+```C
+#if defined(__ARM_FEATURE_DOTPROD)
+(implementation using dotprod instructions)
+#else
+(alternative implementation)
+#endif
+```
+
+A full list of current extensions for Arm can be found [here](https://doc.rust-lang.org/std/arch/macro.is_aarch64_feature_detected.html), while the full list of supported intrinsics is [here](https://doc.rust-lang.org/core/arch/aarch64/index.html).
+
+## Alternative way with Rust using std::simd
+
+You read that there are 2 ways to do SIMD programming with Rust, `std::arch` and `std::simd`, you only saw the first. What about `std::simd`? What would the equivalent program look like using that method?
+
+Here is the same program modified to use `std::simd`. Replace the functions in `average2.rs` with the following:
+
+```Rust
+#[inline(never)]
+fn average_vec(c: &mut [f32], a: &[f32], b: &[f32]) -> () {
+    #[cfg(target_arch = "aarch64")]
```

Comment:
So how does this variant deal with when there's no SIMD support but you are AArch64? as the check is on the callee itself?

Reply:
AArch64 implies ASIMD, unless I'm mistaken? That is, it's not an optional feature of the ISA, like Neon was for Aarch32.

Reply:
AArch64 as an architecture does not imply Adv. SIMD. it's an optional extension.  However the ELF ABI for AArch64 says that on Linux Adv. SIMD is required.

In addition in compilers like GCC you can turn off SIMD support, e.g. `-march=armv8-a+nosimd`, it's not immediately clear to me if you can in Rust.

But the reason I asked was because in this example https://github.com/ArmDeveloperEcosystem/arm-learning-paths/pull/1076/files#diff-090e03bfec44f30dfab5951a79049a53ca2570a7a70849a5af5babca886f18dcR162 you do provide a scalar fall-back. So it wasn't clear to me why you did so in one case but not the other.

Reply:
ah ok, I see. Ok I understand now. Indeed you are correct. I wanted to make a point about having a scalar fall-back, but I didn't think it would be necessary to do that in the rest of the cases, I see why this might cause confusion. I will either provide a scalar fall-back in the rest of the cases, or make a note why there isn't one. Thanks for the suggestion.

Reply:
ok, I added comments in this and the subsequent Rust implementations about scalar/fallback code. If I actually provide the Rust scalar code, the source code will be much larger. I hope that works.